File: sonar-text-plugin/src/test/java/org/sonar/plugins/secrets/rules/AwsAccessKeyIdRuleTest.java
/*
 * SonarQube Text Plugin
 * Copyright (C) 2021-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.plugins.secrets.rules;

import java.util.List;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.groups.Tuple.tuple;
import static org.sonar.plugins.secrets.TestUtils.aNormalizedInputFile;

public class AwsAccessKeyIdRuleTest {

  AwsAccessKeyIdRule underTest = new AwsAccessKeyIdRule();

  @Test
  public void testRuleProperties() {
    assertThat(underTest.getRuleKey()).isEqualTo("S6290");
  }

  @Test
  public void testRuleRegexPositive() {
    List<Secret> secrets = underTest.findSecretsIn(aNormalizedInputFile("public class Foo {\n"
      + "  public static final String KEY = \"AKIAIGKECZXA7AEIJLMQ\"\n"
      + "}"));

    assertThat(secrets)
      .extracting(Secret::getTextRange)
      .extracting("start.line", "start.lineOffset", "end.line", "end.lineOffset")
      .containsOnly(tuple(2, 36, 2, 56));
  }

  @Test
  public void testRuleRegexNegative() {
    List<Secret> secrets = underTest.findSecretsIn(aNormalizedInputFile("public class Foo {\n"
      + "  public static final String KEY = \"AKIGIGKECZXA7AEIJLMQ\"\n"
      + "}"));

    assertThat(secrets).isEmpty();
  }

  @Test
  public void testRuleRegexExamplePositive() {
    List<Secret> secrets = underTest.findSecretsIn(aNormalizedInputFile("public class Foo {\n"
      + "  public static final String KEY = \"AKIAIGKECZXA7EXAMPLF\"\n"
      + "}"));

    assertThat(secrets)
      .extracting(Secret::getTextRange)
      .extracting("start.line", "start.lineOffset", "end.line", "end.lineOffset")
      .containsOnly(tuple(2, 36, 2, 56));
  }

  @Test
  public void testRuleRegexExampleNegative() {
    List<Secret> secrets = underTest.findSecretsIn(aNormalizedInputFile("public class Foo {\n"
      + "  public static final String KEY = \"AKIAIGKECZXA7EXAMPLE\"\n"
      + "}"));

    assertThat(secrets).isEmpty();
  }

}

File: sonar-text-dotnet/src/SonarLint.Secrets.DotNet.UnitTests/Rules/AwsAccessKeyIdRuleTest.cs
/*
 * SonarAnalyzer for Text
 * Copyright (C) 2021-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

// Ported from ...\sonar-secrets-plugin\src\test\java\com\sonarsource\secrets\rules\AwsAccessKeyIdRuleTest.java

using FluentAssertions;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using SonarLint.Secrets.DotNet.Rules;
using SonarLint.VisualStudio.Integration.UnitTests;
using System.Linq;
using static SonarLint.Secrets.DotNet.UnitTests.TestUtils;

namespace SonarLint.Secrets.DotNet.UnitTests.Rules
{
    [TestClass]
    public class AwsAccessKeyIdRuleTest
    {
        [TestMethod]
        public void MefCtor_CheckIsExported()
        {
            MefTestHelpers.CheckTypeCanBeImported<AwsAccessKeyIdRule, ISecretDetector>(null, null);
        }

        [TestMethod]
        public void TestRuleProperties()
        {
            var testSubject = new AwsAccessKeyIdRule();

            testSubject.RuleKey.Should().Be("secrets:S6290");
            testSubject.Name.Should().Be("Amazon Web Services credentials should not be disclosed");
        }

        [TestMethod]
        public void TestRuleRegexPositive()
        {
            var testSubject = new AwsAccessKeyIdRule();

            var input = "public class Foo {\n"
              + "  public static final String KEY = \"AKIAIGKECZXA7AEIJLMQ\"\n"
              + "}";

            var secrets = testSubject.Find(input);

            secrets.Count().Should().Be(1);
            CheckExpectedSecretFound(input, "AKIAIGKECZXA7AEIJLMQ", secrets.First());
            CrossCheckWithJavaResult(input, new (2, 36, 2, 56), secrets.First());
        }

        [TestMethod]
        public void TestRuleRegexNegative()
        {
            var testSubject = new AwsAccessKeyIdRule();

            var secrets = testSubject.Find("public class Foo {\n"
              + "  public static readonly String KEY = \"AKIGIGKECZXA7AEIJLMQ\"\n"
              + "}");

            secrets.Should().BeEmpty();
        }

        [TestMethod]
        public void RestRuleRegexExamplePositive()
        {
            var testSubject = new AwsAccessKeyIdRule();

            var input = "public class Foo {\n"
              + "  public static final String KEY = \"AKIAIGKECZXA7EXAMPLF\"\n"
              + "}";

            var secrets = testSubject.Find(input);

            secrets.Count().Should().Be(1);
            CheckExpectedSecretFound(input, "AKIAIGKECZXA7EXAMPLF", secrets.First());
            CrossCheckWithJavaResult(input, new(2, 36, 2, 56), secrets.First());
        }

        [TestMethod]
        public void TestRuleRegexExampleNegative()
        {
            var testSubject = new AwsAccessKeyIdRule();

            var secrets = testSubject.Find("public class Foo {\n"
              + "  public static readonly String KEY = \"AKIAIGKECZXA7EXAMPLE\"\n"
              + "}");

            secrets.Should().BeEmpty();
        }
    }
}
